File: driver/languages/rust/parts/comparator.rs
use super::data_structures::{ListNode, TreeNode};
use std::rc::Rc;
use std::cell::RefCell;

pub struct Comparator;

impl Comparator {
    pub fn serialize_list_node(head: Option<Box<ListNode>>) -> String {
        let mut vals = Vec::new();
        let mut curr = head;
        while let Some(node) = curr {
            vals.push(node.val);
            curr = node.next;
        }
        format!("{:?}", vals).replace(" ", "")
    }

    pub fn serialize_tree_node(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        if root.is_none() {
            return "[]".to_string();
        }
        let mut res = Vec::new();
        let mut queue = std::collections::VecDeque::new();
        queue.push_back(root);

        while !queue.is_empty() {
            let node_opt = queue.pop_front().unwrap();
            match node_opt {
                Some(node) => {
                    let n = node.borrow();
                    res.push(n.val.to_string());
                    queue.push_back(n.left.clone());
                    queue.push_back(n.right.clone());
                }
                None => {
                    res.push("null".to_string());
                }
            }
        }

        // Trim trailing nulls
        while res.last().map(|s| s.as_str()) == Some("null") {
            res.pop();
        }

        format!("[{}]", res.join(",")).replace(" ", "")
    }

    pub fn serialize<T: std::fmt::Display>(val: T) -> String {
        val.to_string()
    }

    pub fn serialize_vec<T: std::fmt::Display>(vec: Vec<T>) -> String {
        let items: Vec<String> = vec.into_iter().map(|v| v.to_string()).collect();
        format!("[{}]", items.join(",")).replace(" ", "")
    }

    pub fn serialize_string(s: String) -> String {
        format!("\"{}\"", s)
    }

    pub fn serialize_bool(b: bool) -> String {
        b.to_string().to_lowercase()
    }
}

File: driver/languages/go/parts/comparator.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Comparator struct{}

func (c *Comparator) SerializeListNode(head *ListNode) string {
	var vals []string
	curr := head
	for curr != nil {
		vals = append(vals, strconv.Itoa(curr.Val))
		curr = curr.Next
	}
	return "[" + strings.Join(vals, ",") + "]"
}

func (c *Comparator) SerializeTreeNode(root *TreeNode) string {
	if root == nil {
		return "[]"
	}
	var res []string
	queue := []*TreeNode{root}

	for len(queue) > 0 {
		node := queue[0]
		queue = queue[1:]

		if node != nil {
			res = append(res, strconv.Itoa(node.Val))
			queue = append(queue, node.Left)
			queue = append(queue, node.Right)
		} else {
			res = append(res, "null")
		}
	}

	// Trim trailing nulls
	for len(res) > 0 && res[len(res)-1] == "null" {
		res = res[:len(res)-1]
	}

	return "[" + strings.Join(res, ",") + "]"
}

func (c *Comparator) Serialize(val interface{}) string {
	if val == nil {
		return "null"
	}
	switch v := val.(type) {
	case *ListNode:
		return c.SerializeListNode(v)
	case *TreeNode:
		return c.SerializeTreeNode(v)
	case string:
		return "\"" + v + "\""
	case bool:
		if v {
			return "true"
		}
		return "false"
	case []int:
		var strs []string
		for _, x := range v {
			strs = append(strs, strconv.Itoa(x))
		}
		return "[" + strings.Join(strs, ",") + "]"
	case []string:
		var strs []string
		for _, x := range v {
			strs = append(strs, "\""+x+"\"")
		}
		return "[" + strings.Join(strs, ",") + "]"
	case [][]int:
		var rows []string
		for _, row := range v {
			var strs []string
			for _, x := range row {
				strs = append(strs, strconv.Itoa(x))
			}
			rows = append(rows, "["+strings.Join(strs, ",")+"]")
		}
		return "[" + strings.Join(rows, ",") + "]"
	case []interface{}:
	    var strs []string
	    for _, x := range v {
	        strs = append(strs, c.Serialize(x))
	    }
	    return "[" + strings.Join(strs, ",") + "]"
	default:
		return fmt.Sprintf("%v", v)
	}
}
